OpenRA.Mods.Common/Projectiles/Railgun.cs

```diff
@@ -0,0 +1,246 @@
+		public readonly int HelixPitch = 512;
+
+		[Desc("Helix radius gets + this value per tick during drawing")]
+		public readonly int HelixRadiusDeltaPerTick = 8;
+
+		[Desc("Helix alpha gets + this value per tick during drawing; hence negative value makes it fade over time.")]
+		public readonly int HelixAlphaDeltaPerTick = -8;
+
+		[Desc("Helix spins by this much over time each tick. (1 = 360/1024 degrees)")]
+		public readonly int HelixAngleDeltaPerTick = 16;
+
+		[Desc("Draw each cycle of helix with this many quantization steps")]
+		public readonly int QuantizationCount = 16;
+
+		[Desc("Helix color in (A),R,G,B.")]
+		public readonly Color HelixColor = Color.FromArgb(128, 255, 255, 255);
+
+		[Desc("Draw helix in PlayerColor? Overrides RGB part of the HelixColor. (Still uses HelixColor's alpha information)")]
+		public readonly bool HelixPlayerColor = false;
+
+		[Desc("Impact animation.")]
+		public readonly string HitAnim = null;
+
+		[Desc("Sequence of impact animation to use.")]
+		[SequenceReference("HitAnim")]
+		public readonly string HitAnimSequence = "idle";
+
+		[PaletteReference]
+		public readonly string HitAnimPalette = "effect";
+
+		[Desc("Scan radius for actors damaged by beam. If set to zero (default), it will automatically scale to the largest health shape.",
+			"Only set custom values if you know what you're doing.")]
+		public WDist AreaVictimScanRadius = WDist.Zero;
+
+		[Desc("Scan radius for actors with projectile-blocking trait. If set to zero (default), it will automatically scale",
+			"to the blocker with the largest health shape. Only set custom values if you know what you're doing.")]
+		public WDist BlockerScanRadius = WDist.Zero;
+
+		public IProjectile Create(ProjectileArgs args)
+		{
+			var bc = BeamPlayerColor ? Color.FromArgb(BeamColor.A, args.SourceActor.Owner.Color.RGB) : BeamColor;
+			var hc = HelixPlayerColor ? Color.FromArgb(HelixColor.A, args.SourceActor.Owner.Color.RGB) : HelixColor;
+			return new Railgun(args, this, bc, hc);
+		}
+
+		public void RulesetLoaded(Ruleset rules, WeaponInfo wi)
+		{
+			if (BlockerScanRadius == WDist.Zero)
+				BlockerScanRadius = Util.MinimumRequiredBlockerScanRadius(rules);
+
+			if (AreaVictimScanRadius == WDist.Zero)
+				AreaVictimScanRadius = Util.MinimumRequiredVictimScanRadius(rules);
+		}
+	}
+
+	public class Railgun : IProjectile
+	{
+		readonly ProjectileArgs args;
+		readonly RailgunInfo info;
+		readonly Animation hitanim;
+		public readonly Color BeamColor;
+		public readonly Color HelixColor;
+
+		int ticks = 0;
+		bool animationComplete;
+		WPos target;
+
+		// Computing these in Railgun instead of RailgunRenderable saves Info.Duration ticks of computation.
+		// Fortunately, railguns don't track the target.
+		int cycleCount;
+		WVec sourceToTarget;
+		WVec forwardStep;
+		WVec leftVector;
+		WVec upVector;
+		WAngle angleStep;
+
+		public int CycleCount { get { return cycleCount; } }
+		public WVec SourceToTarget { get { return sourceToTarget; } }
```

Comment:
Style nit: declare these as `{ get; private set; }` and the compiler will define the backing fields for you, so you can remove the variables defined just above.